components/raftstore-v2/src/raft/peer.rs

```diff
@@ -849,4 +859,131 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
             Err(e) => slog_panic!(self.logger, "failed to load term"; "index" => idx, "err" => ?e),
         }
     }
+
+    #[inline]
+    pub fn peer_status_context_mut(&mut self) -> &mut PeerStatusContext {
+        &mut self.peer_status_context
+    }
+
+    #[inline]
+    pub fn peer_status_context(&self) -> &PeerStatusContext {
+        &self.peer_status_context
+    }
+
+    /// Returns `true` if any peer recover from connectivity problem.
+    ///
+    /// A peer can become pending or down if it has not responded for a
+    /// long time. If it becomes normal again, PD need to be notified.
+    pub fn any_new_peer_catch_up(&mut self, peer_id: u64) -> bool {
+        if self.peer_status_context.is_empty() {
+            return false;
+        }
+        if !self.is_leader() {
+            self.peer_status_context.reset();
+            return false;
+        }
+
+        if self.peer_status_context.has_down(peer_id) {
+            return true;
+        }
+
+        let logger = self.logger.clone();
+        let leader_id = self.peer_id();
+        let region_id = self.region_id();
+        self.peer_status_context
+            .retain_pendings(|(peer_id, pending_after)| {
+                // TODO check wait data peers here
+                let truncated_idx = self.raft_group.store().entry_storage().truncated_index();
+                if let Some(progress) = self.raft_group.raft.prs().get(*peer_id) {
+                    if progress.matched >= truncated_idx {
+                        let elapsed = duration_to_sec(pending_after.saturating_elapsed());
+                        RAFT_PEER_PENDING_DURATION.observe(elapsed);
+                        debug!(
+                            logger,
+                            "peer has caught up logs";
+                            "region_id" => %region_id,
```

Comment:
logger already has "region_id" and "peer_id". And "peer_id" usually means the peer that is processing instead of being processed.

Reply:
I'll remove the region_id and leader_id and add a `from_peer_id` instead.